### PR TITLE
add types for kap plugins

### PR DESCRIPTION
**Pull request by connor4312**
This PR adds typings for writing [kap plugins](https://github.com/wulkano/kap/blob/master/docs/plugins.md).

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types
- [x] This is not an npm packages and does not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/kap-plugin/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    defaultFileName: string;
+    filePath(): Promise<string>;
+    config: ConfigStore<T & BuiltInConfig>;
+    request: typeof got;
+    copyToClipboard(text: string): void;
+    notify(text: string, action?: () => void): void;
+    setProgress(text: string, percentage: number): void;
+    openConfigFile(): void;
+    cancel(): void;
+    waitForDeepLink(): Promise<string>;
```
Comment (sindresorhus) [inline types/kap-plugin/index.d.ts]: Can you include some doc comments taken from the docs [here](https://github.com/wulkano/kap/blob/master/docs/plugins.md#action)?

File: types/kap-plugin/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": [
+            "es6"
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
```
Comment (sindresorhus) [inline types/kap-plugin/tsconfig.json]: Not sure how DT works, but would be nice to make everything strict with `"strict": true`.

File: types/kap-plugin/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// Project: https://github.com/wulkano/kap/blob/master/docs/plugins.md
+// Definitions by: Connor Peet <https://github.com/connor4312>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.0
+
+import * as got from 'got';
+import { JSONSchema7 } from 'json-schema';
+
+export interface BuiltInConfig {
+    accessToken?: string;
```
Comment (sindresorhus) [inline types/kap-plugin/index.d.ts]: This is not something Kap has.

File: types/kap-plugin/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+export interface ConfigStore<T> {
+    get<K extends keyof T>(key: K): T[K];
+}
+
+export interface KapContext<T> {
+    format: string;
```
Comment (sindresorhus) [inline types/kap-plugin/index.d.ts]: Missing `prettyFormat` below `format`.

File: types/kap-plugin/index.d.ts (hunk 3 above)
Comment (connor4312) [reply types/kap-plugin/index.d.ts]: Okay, I was looking at the [description of OAuth](https://github.com/wulkano/kap/blob/master/docs/plugins.md#oauth). What's the right way to get the access token?

Comment (sindresorhus) [reply types/kap-plugin/index.d.ts]: It’s up to you to fetch it and store it.

File: types/kap-plugin/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    defaultFileName: string;
+    filePath(): Promise<string>;
+    config: ConfigStore<T & BuiltInConfig>;
```
[the remark this answers is not in the document]
Comment (connor4312) [reply types/kap-plugin/index.d.ts]: It looks like DT compiles with TS versions (down to 3.0 for this module) where the `strict` shorthand isn't supported.

File: types/kap-plugin/index.d.ts (hunk 3 above)
Comment (connor4312) [reply types/kap-plugin/index.d.ts]: Ah, I see; it describes a recommended list of steps, not an enforced one, got it 🙂

File: types/kap-plugin/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+     * the Kap export window. If the cancelation was not the result of a user
+     * gesture, use .notify() to inform the user why it was canceled.
+     */
+    cancel(): void;
+
+    /**
+     * Returns a Promise that resolves when a deep link for this plugin is
+     * opened. The link should be in the format `kap://plugins/{pluginName}/{rest}`,
+     * where pluginName is the npm package name and rest is the string the
+     * Promise will resolve with. This is useful for OAuth flows.
```
Comment (sindresorhus) [inline types/kap-plugin/index.d.ts]:
```suggestion
     * Promise will resolve with. This is useful for [OAuth flows](https://github.com/wulkano/kap/blob/master/docs/plugins.md#oauth).
```

**Review by sindresorhus — APPROVED** [review]

**Pull request merged**